### PR TITLE
Adding support for Not

### LINKED ISSUE #2: Add support for NotAction

### LINKED ISSUE #3: Add support for NotPrincipal

### LINKED ISSUE #4: Add support for NotResource

### PR DESCRIPTION
**Pull request by ringods**
Fixes #2 - Add support for NotAction
Fixes #3 - Add support for NotPrincipal
Fixes #4 - Add support for NotResource

### CODE REVIEW

**Comment by tdpauw** [comment]
Wow, thank you @ringods for this pull request. This is an amazing contribution!

**Comment by tdpauw** [comment]
Wow, thank you @ringods for this pull request. This is an amazing contribution!

File: src/statement/statement.ts [hunk 1]
```diff
@@ -65,13 +71,21 @@ export class Statement {
...
   }
 
+  validateForNotResourcePolicy() {
+    const errors = this.validateForAnyPolicy();
+    if (Object.keys(this.principals).length === 0) {
+      errors.push(`Statement(${this.sid}) must specify at least one IAM principal.`);
+    }
+    return errors;
+  }
+
```
Comment (tdpauw) [inline src/statement/statement.ts:81]: Why was this method added? I think this is not necessary.

`validateForResourcePolicy` validates a statement in the context of a Policy attached to an AWS Resource like an S3 bucket, a KMS Key, a SecretsManager secret, ...
Likewise `validateForIdentityPolicy` validates a statement in the context of a policy attached to an AWS Identity like an IAM User, Group, or Role.

The `ForResource` has nothing to do with `statement.resources`. Therefore no need to add a `ForNotResource` variant.

File: tests/statement/statement.spec.ts [hunk 2]
```diff
@@ -76,7 +76,31 @@ describe('#Statement', function() {
...
+    });
+
+    it('should be invalid for identity-based policies', function() {
+      const errors = statement.validateForIdentityPolicy();
+      expect(errors).to.have.length(2);
+      expect(errors[0]).to.equal(
+          'Statement(NotValidForResource) cannot specify any IAM principals.',
+      );
+      expect(errors[1]).to.equal(
+          'Statement(NotValidForResource) must specify at least one resource or notresource.',
```
Comment (tdpauw) [inline tests/statement/statement.spec.ts:103]: I think these tests are not required as the function `validateForNoteResourcePolicy` is also not needed.

However, we should add a ValidForIdentity case where the statement has a `NotResource`.

File: src/statement/statement.ts (hunk 1 above)
Comment (ringods) [reply src/statement/statement.ts]: Got it. Removed the method.

File: tests/statement/statement.spec.ts (hunk 2 above)
Comment (ringods) [reply tests/statement/statement.spec.ts]: Got it. Removed the tests.

**Comment by ringods** [comment]
On `PolicyDocument`, I split the functionality of `.json` accessor in two: first retrieve the intermediate JS object structure using `.object` before invoking `JSON.stringify`. In our use case, we need to embed a policy document within a piece of Cloudformation for AWS Organisations. We may only run `JSON.stringify` once at the end. Having the `.object` accessor allows us to embed easier in a bigger JS object structure.

**Comment by tdpauw** [comment]
Yes, I love the idea. Separation of concerns for the win!

**Review by tdpauw — CHANGES_REQUESTED** [review]
Looking good! Just one small missing piece.

`validateForResourcePolicy` and `validateForIdentityPolicy` should both also check for `this.notprincipals`

```typescript
validateForResourcePolicy() {
    const errors = this.validateForAnyPolicy();
    if (Object.keys(this.principals).length === 0 && Object.keys(this.notPrincipal).length === 0) {
      errors.push(`Statement(${this.sid}) must specify at least one IAM principal.`);
    }
    return errors;
  }
```

```typescript
validateForIdentityPolicy() {
    const errors = this.validateForAnyPolicy();
     if (Object.keys(this.principals).length > 0 || Object.keys(this.notPrincipal).length > 0) {
       errors.push(`Statement(${this.sid}) cannot specify any IAM principals.`);
     }
    ...
}
```

**Review by tdpauw — APPROVED** [review]

**Pull request merged**